dedupe/training.py

```diff
@@ -138,9 +139,15 @@ def score(predicate: Predicate) -> float:
 
         return candidates
 
-    def cover(self, pairs: TrainingExamples) -> Cover:
+    def cover(self, pairs: TrainingExamples, index_predicates: bool = True) -> Cover:
```

Comment:
don't love this change. is it really ugly to avoid this?

Reply:
Have you seen the other changes from this commit? and the corresponding commit message? I think the current way that we deal with this, where in DisagreementLearner.learn_predicates() we do `self.blocker.block_learner.blocker.predicates.copy()` is a terrible code smell, reaching 3 levels deep, where `blocker.block_learner.blocker` are all mean the same thing. So I wanted to actually create public APIs at each level.

The other problem there is the pattern of "temporarily delete some predicates, do some stuff, and then restore them" which I didn't like.

I agree that this commit does add more lines of code, which I don't like, but what I like less was the unofficial and haphazard methods we used.

When you say "this change", do you mean "adding index_predicates as an arg to this function"? We could probably adjust this. Or do you mean the larger-scale changes I mention above?

Reply:
those are all really good arguments. for me, the ideal thing would be for the `BlockLearner` classes to only have two public methods, the `__init__` and the `learn` methods. in particular, my initial instinct is that `cover` should not be called publicly.

i completely agree that we are violating that with how we are reaching into the objects with `self.blocker.block_learner.blocker.predicates.copy()`

i really like the rest of the changes in this PR. could we go back to the `self.blocker.block_learner.blocker.predicates.copy()` pattern for now, and open up separate issue for a design for that?

Reply:
actually, i'm good with this!

Reply:
i'm really struggling with this. i think this is better than the code that it is replacing, but it seems like this filtering should not be the responsibility of the BlockLearner class.

Reply:
hmm... what about this design.

```python
class BlockLearner(ABC):
    def learn(self, matches, recall, candidate_types='simple', selected_predicates=None):
        comparison_cover = self.comparison_cover
        if selected_predicates:
            candidate_preds = comparison_cover & selected_predicates
        else:
            candidate_preds = comparison_cover.keys()
   
        match_cover = self.cover(matches, candidate_preds)
        ...



    def cover(self, pairs, candidate_preds) -> Cover:
        predicate_cover = {}
        for predicate in candidate_preds:  # type: ignore
        ...
        
```

then this labeler code could just be

```python
    def learn_predicates(
        self, recall: float, selected_predicates
    ) -> tuple[Predicate, ...]:

        learned_preds: tuple[Predicate, ...]
        dupes = [pair for label, pair in zip(self.y, self.pairs) if label]
        learned_preds = self.blocker.block_learner.learn(
                dupes, recall=recall, candidate_types="random forest",
                selected_predicates
            )
```

and so then it would ultimately be the responsible for the `train` method in the api.py
 
to filter the predicates based and pass them on appropriately.

Reply:
I'm trying to wrap my head around the general pattern of what we're doing, correct me if I'm wrong. Right now there are three sets of predicates:
1. all the available predicates from the data model
2. A subset of 1, all the predicates that the BlockLearners use as candidates.
3. A subset of 2, the final predicates chosen by the BlockLearners that we use to actually do the blocking in Dedupe.pairs().

Right now BlockLearners are handed 1, and it filters that down to 2 using whether or not predicates are index predicates or not, using the `index_predicates` flag. Then it does the step of filtering from 2 to 3.

I agree that this is bad, these concerns should be separated. The filtering from 1 to 2 should happen external to the BlockLearners. The block learners should just get handed set 2, and they select set 3 from that. That is the direction that https://github.com/dedupeio/dedupe/issues/1052 is trying to go in.

I'm not sure the 1->2 filtering should happen in train() though. Can't we do this at the very beginning, when the data model is first given to Dedupe? Or at least in training.DedupeBlockLearner.__init__, because there we call 

```
self.blocker = blocking.Fingerprinter(predicates)
self.blocker.index_all(data)
```

which makes us index ALL the predicates, even if then in subsequent calls to learn() or cover() we then filter out the index predicates and don't use them.

If you think this is at least an improvement, how about you merge it and then we can actually write PRs that go in this even better direction, and we have something concrete to talk about? I'm finding it hard without actually testing things out, but I don't want to have merge conflicts later so I want main to be stable.

Reply:
But yes I think your suggestion is better than what I have currently.

Reply:
This conversation is getting too big, should be it's own issue?

Reply:
@fgregg in case you didn't see this